gapic/schema/api.py

```diff
@@ -560,6 +561,133 @@ def mixin_http_options(self):
             res[s] = [rule for rule in opt_gen if rule]
         return res
 
+    def get_auto_populated_fields_candidates(
+        self,
+        message: MethodDescriptorProto,
+    ) -> Sequence[str]:
+        """Return fields which meet the criteria to be automatically populated for a given message.
+        For a field to be automatically populated, all the below configurations must
+        be true according to https://google.aip.dev/client-libraries/4235:
+
+        - The field must be of type string
+        - The field must be at the top-level of the request message
+        - The field must not be annotated with google.api.field_behavior = REQUIRED.
+        - The field must be annotated with google.api.field_info.format = UUID4.
+
+        Args:
+            message[MethodDescriptorProto]: Describes a method of a service.
+        Return:
+            Sequence[str]: A list of candidates to be automatically populated
+                according to https://google.aip.dev/client-libraries/4235.
+        """
+        top_level_request_message = self.messages[
+            message.input_type.lstrip(".")
+        ]
+        return [
+            field.name
+            for field in top_level_request_message.fields.values()
+            if field.type == wrappers.PrimitiveType.build(str)
+            and not field.required
+            and field.uuid4
+        ]
+
+    def get_auto_populated_fields(
+        self,
+        method_settings: client_pb2.MethodSettings,
+    ) -> Sequence[str]:
+        """
+        Return a list of auto-populated fields based on the service YAML.
+
+        Args:
+            method_settings (client_pb2.MethodSettings): The settings
+            to be used when generating API methods.
+        Return:
+            Sequence[str]: A sequence of all fields to be automatically
+                populated, taking into account the requirements
+                from AIP-4235.
+                https://google.aip.dev/client-libraries/4235
+
+        Raises:
+            ValueError: if `auto_populated_fields` are incorrectly set in the service config
+        """
+        auto_populated_fields = []
+
+        def _check_service_yaml_validity(method_in_scope: MethodDescriptorProto, method_settings: client_pb2.MethodSettings):
```

Comment:
What's the advantage of having this as a nested function? It's not terribly short, so breaking it out and clarifying the name might its purpose clearer.

Reply:
Fixed in c3563c18660e4bef34f62381f91ba22f0e42f5a8